lib/pbench/server/database/models/datasets.py

```diff
@@ -436,11 +435,11 @@ class Dataset(Database.Base):
     # Dataset name
     name = Column(String(255), unique=False, nullable=False)
 
-    # ID of the owning user
-    owner_id = Column(Integer, ForeignKey("users.id"), nullable=False)
+    # Keycloak ID of the owning user
+    owner_id = Column(String(255), nullable=True)
 
-    # Indirect reference to the owning User record
-    owner = relationship("User")
+    # username of the owning user
+    owner = Column(String(255), nullable=True)
```

Comment:
I'm not sure we really want to try to keep this "redundant" relationship.

Does Keycloak allow renaming an existing user, for example? If it does, we risk retaining stale name strings in addition to the unique id. In any case, this doesn't directly translate from the previous relationship, where `owner` was the full `User` join providing all the user metadata (e.g., role). We'll still have to call Keycloak for any of that, and I'm not sure that special-casing "username" is worthwhile.

Reply:
> (This might be encoded in the token, or we might have to ask Keycloak for it.)

Yes, I think both are encoded in the token, but the "token" isn't the owner of a dataset, just the user's UUID, so we don't have access to that in this context. (We may or may not have access to the token of the *requestor*, but in general we won't have a token for the *owner*.)

> so I think we will want to cache the user name for that purpose...so, I think we'll want this column.

I'm strongly against that model. Caching external values is messy; I don't want us to own the synchronization issues. And even if we did cache, we shouldn't be doing it by scattering ID and name across various unrelated tables: we'd do it with a single centralized cache repository (whether a SQL table, a Redis key, or whatever) that can be easily and directly managed.

Reply:
The problem is that if the user changes his/her username then we are left with wrong and/or stale user metadata (username) related to the dataset. So we decided to keep just the user UUID for now as the only user-related field since that is guaranteed to be unique. If we ever have to look up the user we can make a request to the Keycloak to give us the user information based on the user uuid.

Reply:
> I think both are encoded in the token, but the "token" isn't the owner of a dataset, just the user's UUID, so we don't have access to that in this context. (We may or may not have access to the token of the _requestor_, but in general we won't have a token for the _owner_.)

The user is the owner of the dataset; the user is the requestor (when it's an authenticated request).  The user is identified by the UUID which the Server extracts from the supplied token.  These parts are easy.

The challenging part is when a user's request needs to refer to _some other_ user.

> > so I think we will want to cache the user name for that purpose...so, I think we'll want this column.
> 
> I'm strongly against that model. Caching external values is messy; I don't want us to own the synchronization issues.

@dbutenhof, I think you are placing too high an expectation on this.  The Pbench Server translation of a given user's UUID to the user's name does not need to reflect information which is any newer than the last time that user logged into the Pbench Server (IMO).  That is, each time a user logs into the Pbench Server, we can update our cached value of the user's name, and then we can continue to use that value until the next time that user logs in.  I think that will be perfectly adequate.  I don't see sufficient benefit in trying to update the cache any more often than that (for all the reasons you reference).  (And, if the user doesn't log into the Pbench Server after changing their username on the Keycloak server, then it doesn't matter 'cause, in our reality, that user still has the old name. 🙂)

> if we did cache, we shouldn't be doing it by scattering ID and name across various unrelated tables

Agreed.  We should have it in a single place which either we can look up (e.g., with the UUID) or which has a relationship to the record in question.

> The problem is that if the user changes his/her username then we are left with wrong and/or stale user metadata (username) related to the dataset.

Which is why, @npalaska, I'm pushing to use the UUID, not the username, internally to the Pbench Server.

> If we ever have to look up the user we can make a request to the Keycloak to give us the user information based on the user uuid.

It's good to have you confirm that we can do that, but I suspect that we are going to want user information too often to want to go to that trouble.  (E.g., if the client does a search of public results, presumably we are going to want the Dashboard to display ownership information with those results, and, presumably, we're going to want to present something friendlier than UUIDs; so if the query comes back with 1000 results, we're probably not going to want to hit the Keycloak server 1000 times for that one client request....)

Reply:
> (And, if the user doesn't log into the Pbench Server after changing their username on the Keycloak server, then it doesn't matter 'cause, in our reality, that user still has the old name. )

Disagree. A years-old dataset is owned by someone who's changed username, and someone (with access, possibly an admin, or they're public datasets) searches for datasets owned by that user... using the new name they know. We need to find that dataset, regardless of our stale cache... not make them guess what the username might have been in the past.